test-app/src/main/java/org/readium/r2/testapp/reader/tts/TtsControls.kt

```diff
@@ -0,0 +1,198 @@
+/*
+ * Copyright 2021 Readium Foundation. All rights reserved.
+ * Use of this source code is governed by the BSD-style license
+ * available in the top-level LICENSE file of the project.
+ */
+
+package org.readium.r2.testapp.reader.tts
+
+import androidx.compose.foundation.layout.*
+import androidx.compose.material.*
+import androidx.compose.material.icons.Icons
+import androidx.compose.material.icons.filled.*
+import androidx.compose.runtime.*
+import androidx.compose.ui.Alignment
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.platform.LocalContext
+import androidx.compose.ui.res.stringResource
+import androidx.compose.ui.unit.dp
+import org.readium.r2.navigator.tts.PublicationSpeechSynthesizer.Configuration
+import org.readium.r2.navigator.tts.TtsEngine.Voice
+import org.readium.r2.shared.ExperimentalReadiumApi
+import org.readium.r2.shared.util.Language
+import org.readium.r2.testapp.R
+import org.readium.r2.testapp.shared.views.SelectorListItem
+import org.readium.r2.testapp.utils.extensions.asStateWhenStarted
+import java.text.DecimalFormat
+
+/**
+ * TTS controls bar displayed at the bottom of the screen when speaking a publication.
+ */
+@OptIn(ExperimentalReadiumApi::class)
+@Composable
+fun TtsControls(model: TtsViewModel, modifier: Modifier = Modifier) {
+    val showControls by model.state.asStateWhenStarted { it.showControls }
+    val isPlaying by model.state.asStateWhenStarted { it.isPlaying }
+    val settings by model.state.asStateWhenStarted { it.settings }
```

Comment:
I started with the usual:

```kotlin
val state by model.state.flowWithLocalLifecycle().collectAsState()
```

But as I used a single `State` object that is updated every time a word is spoken, that triggered a lot of recomposition. I came up with this `asStateWhenStarted` extension that select a property of the global state instead, and is basically (it's a bit more complicated, see `Flow.kt`) equivalent to:

```kotlin
val showControls by remember {
    model.state
        .map { it.showControls }
        .flowWithLocalLifecycle()
}.collectAsState()
```